dm-dedup-rw.c

```diff
@@ -64,15 +66,19 @@ static int merge_data(struct dedup_config *dc, struct page *page,
 		goto out;
 	}
 
-	src_page_vaddr = page_address(bio->bi_io_vec->bv_page);
-	dest_page_vaddr = page_address(page);
-
-	src_page_vaddr = src_page_vaddr + bio->bi_io_vec->bv_offset;
 	/* Locating the right sector to merge */
-	dest_page_vaddr = dest_page_vaddr + (to_bytes(position));
+	dest_page_vaddr = page_address(page) + to_bytes(position);
+
+	bio_for_each_segment(bvec, bio, iter) {
+		src_page_vaddr = bio_data(bio);
+
+		/* Merging Data */
+		memmove(dest_page_vaddr, src_page_vaddr, bio_cur_bytes(bio));
+
+		/* Updating destination address */
+		dest_page_vaddr += bio_cur_bytes(bio);
+    }
```

Comment:
Yes, bio_data and bio_cur_bytes get the idx from the bio->bi_iter. However, bio_for_each_segment advances *not* bio->bi_iter but the iter which is passed as the third argument to bio_for_each_segment.

Look, e.g., at bio_segments() function. If your logic would be correct, then the side effect of this function calling bio_for_each_segment() would be that the bio->bi_iter is advanced...